### ISSUE TITLE
More flexible Resource cache eviction policy

### ISSUE THREAD

**Issue by pcantrell** [bustoutsolutions/siesta#31]
Resources are currently eligible for deallocation on a low memory event. This works fine in most iOS apps, but (1) will not work on desktop apps, where low memory events do not exist, and (2) may cause memory churn for apps cycling rapidly through many resources.

`WeakCache` should support a more flexible eviction policy — perhaps akin to `NSCache`, perhaps more rule-based and resource-aware.

**Comment by onekiloparsec** [comment]
That's definitely something we must work on for the desktop, I agree. NSCache has 2 properties we could build upon, maybe? 

- var totalCostLimit: Int, The maximum total cost that the cache can hold before it starts evicting objects
- var countLimit: Int, The maximum number of objects the cache should hold

**Comment by pcantrell** [comment]
I decided not to implement `WeakCache` using `NSCache` because of the complexity of the latter. As Matt Thompson put it, it’s “kind of a hot mess.” In particular, I was concerned about making Siesta’s strong → weak transition work in terms of `NSDiscardableContent`. The trick is that `WeakCache` needs to be able to to make refs go weak when the cache _wants_ to discard them, but not actually _remove_ them from the cache until the weak ref goes nil. I investigated a while, and it looked like it might conceivably be possible to make `NSCache` do this, but it would have to rely on dubious assumptions about the exact sequence of calls and internal behavior. Thus no `NSCache.`
